# WeRateDogs Data Wrangling

# Gather

In [6]:
#import required libraries
import pandas as pd
import os
import requests
import tweepy
import json

## The WeRateDogs Twitter archive

In [2]:
df_archive = pd.read_csv('twitter-archive-enhanced.csv')
df_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

## The Tweet image predictions
what breed of dog (or other object, animal, etc.) is present in each tweet according to a neural network.

In [8]:
# Download image predictions file if it doesn't exist
pred_url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
file_name = pred_url.split('/')[-1]

if os.path.exists(file_name):
    print(f"File {file_name} already exists")
else:
    r = requests.get(pred_url)
    # write the reponse to a local file
    with open('image-predictions.tsv','wb') as file:
        file.write(r.content)
        print(f"successfully downloaded {file_name} and written to disc")

File image-predictions.tsv already exists


In [4]:
# Read the image predictions file into a DataFrame
df_pred = pd.read_csv('image-predictions.tsv',sep='\t')
df_pred.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

## Twitter API
Each tweet's retweet count and favorite ("like") count at minimum, and any additional data you find interesting. 

Using the tweet IDs in the WeRateDogs Twitter archive, query the Twitter API for each tweet's JSON data using Python's Tweepy library and store each tweet's entire set of JSON data in a file called tweet_json.txt file. Each tweet's JSON data should be written to its own line. 

Then read this .txt file line by line into a pandas DataFrame with (at minimum) tweet ID, retweet count, and favorite count. 

Note: do not include your Twitter API keys, secrets, and tokens in your project submission.

In [15]:
tweet_ids = df_archive['tweet_id']
json_file = 'tweet_json.txt'


# only Query the API if the tweet_json.txt file doesn't exist
if os.path.exists(json_file):
    print(f"File {json_file} already exists")
else:
    # import the keys files where the API keys are stored and it is not submitted with the project
    import keys
    # Twitter API authorization 
    auth = tweepy.OAuthHandler(keys.consumer_key, keys.consumer_secret)
    auth.set_access_token(keys.access_token, keys.access_secret)

    api = tweepy.API(auth)
    
    count = 0
    fails_dict = {}
    # Save each tweet's returned JSON as a new line in a .txt file
    with open(json_file, 'w') as outfile:
    # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
        for tweet_id in tweet_ids:
            count += 1
            print(str(count) + ": " + str(tweet_id))
            try:
                tweet = api.get_status(tweet_id, tweet_mode='extended')
                print("Success")
                json.dump(tweet._json, outfile)
                outfile.write('\n')
            except tweepy.TweepError as e:
                print("Fail")
                fails_dict[tweet_id] = e
                pass
    print(f"File {json_file} successfully written to Disk")
    print("Some tweets returned the following errors")
    print(fails_dict)



File tweet_json.txt already exists


In [20]:
# read .txt file line by line into a pandas DataFrame

labels = ['tweet_id','retweet_count','favorite_count']

df_api = pd.DataFrame(columns=labels)

with open(json_file) as file:
    for line in file:
        data = json.loads(line)
        df_api = df_api.append({
            'tweet_id':data['id'],
            'retweet_count':data['retweet_count'],
            'favorite_count':data['favorite_count']
            },ignore_index=True)
        
df_api.sample(5)

tweet_id retweet_count favorite_count
1840  675354435921575936         15810          30900
430   819015337530290176         36454              0
1779  676946864479084545           346           1660
117   869227993411051520          3398          18968
176   856526610513747968          1755          11260

# Assess

## Twitter Archive

In [21]:
df_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [53]:
df_archive.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
794   773336787167145985                    NaN                  NaN   
2316  666649482315059201                    NaN                  NaN   
1141  727644517743104000                    NaN                  NaN   
1415  698703483621523456                    NaN                  NaN   
1811  676819651066732545                    NaN                  NaN   

                      timestamp  \
794   2016-09-07 01:47:12 +0000   
2316  2015-11-17 16:09:56 +0000   
1141  2016-05-03 23:42:26 +0000   
1415  2016-02-14 03:01:06 +0000   
1811  2015-12-15 17:42:34 +0000   

                                                 source  \
794   <a href="http://twitter.com/download/iphone" r...   
2316  <a href="http://twitter.com/download/iphone" r...   
1141  <a href="http://twitter.com/download/iphone" r...   
1415  <a href="http://twitter.com/download/iphone" r...   
1811  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
794   RT @dog_rates: Meet Fizz. She thinks love is a...         7.713808e+17   
2316  Cool dog. Enjoys couch. Low monotone bark. Ver...                  NaN   
1141  Here's a doggo struggling to cope with the win...                  NaN   
1415  This is Rusty. He has no respect for POULTRY p...                  NaN   
1811  Watch out Airbud. This pupper is also good at ...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
794               4.196984e+09  2016-09-01 16:14:48 +0000   
2316                       NaN                        NaN   
1141                       NaN                        NaN   
1415                       NaN                        NaN   
1811                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
794   https://twitter.com/dog_rates/status/771380798...                11   
2316  https://twitter.com/dog_rates/status/666649482...                 4   
1141  https://twitter.com/dog_rates/status/727644517...                13   
1415  https://twitter.com/dog_rates/status/698703483...                 7   
1811  https://twitter.com/dog_rates/status/676819651...                12   

      rating_denominator   name  doggo floofer  pupper puppo  
794                   10   Fizz   None    None    None  None  
2316                  10   None   None    None    None  None  
1141                  10   None  doggo    None    None  None  
1415                  10  Rusty   None    None    None  None  
1811                  10   None   None    None  pupper  None

In [24]:
# check for duplicated entries
sum(df_archive['tweet_id'].duplicated())

0

In [28]:
df_archive['rating_denominator'].value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [102]:
df_archive['rating_numerator'].value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [32]:
df_archive['name'].value_counts()

None       745
a           55
Charlie     12
Cooper      11
Lucy        11
          ... 
Miguel       1
Blu          1
Strider      1
Barney       1
Bobb         1
Name: name, Length: 957, dtype: int64

In [44]:
df_archive['doggo'].value_counts()

None     2259
doggo      97
Name: doggo, dtype: int64

In [270]:
df_archive[['doggo','floofer','pupper','puppo']].value_counts()

doggo  floofer  pupper  puppo
None   None     None    None     1976
                pupper  None      245
doggo  None     None    None       83
None   None     None    puppo      29
doggo  None     pupper  None       12
None   floofer  None    None        9
doggo  floofer  None    None        1
       None     None    puppo       1
dtype: int64

In [45]:
df_archive['in_reply_to_status_id'].value_counts()

6.671522e+17    2
8.562860e+17    1
8.131273e+17    1
6.754971e+17    1
6.827884e+17    1
               ..
8.482121e+17    1
6.715449e+17    1
6.936422e+17    1
6.849598e+17    1
7.331095e+17    1
Name: in_reply_to_status_id, Length: 77, dtype: int64

In [25]:
df_archive['source'].value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

### Quality

* missing values in the columns :'in_reply_to_status_id','in_reply_to_user_id ','retweeted_status_id', 'retweeted_status_user_id','retweeted_status_timestamp','expanded_urls'
* 'timestamp' and 'retweeted_status_timestamp' column is has type of srting. should be datetime.
* 'name' column has "None"  and "a" values.
* 'source' column is the complete a tag.
* 'expanded_urls' have missing values.
* the 'rating_denominator' column is not always 10.
* the 'rating_numerator' has inconsistent valus.. 
* 'in_reply_to_status_id','in_reply_to_user_id ','retweeted_status_id', 'retweeted_status_user_id' columns are of type float and using the sceintific notation. should be integer.
* 181 entries are retweets.
* 78 entries are relies.
* 12 entries have both lables 'doggo' and 'pupper'
* 1 entry have both labels 'doggo' and 'floofer

### Tidiness
* The dog stages columns ('doggo', 'flooofer', 'pupper', 'puppo') should be one column instead.

## Tweet Image Predictions

In [312]:
df_pred.sample(5)

tweet_id                                          jpg_url  \
540   676975532580409345  https://pbs.twimg.com/media/CWUZpydWcAAeipD.jpg   
25    666362758909284353  https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
1465  778624900596654080  https://pbs.twimg.com/media/Cs47N3eWcAEmgiW.jpg   
649   681891461017812993  https://pbs.twimg.com/media/CXaQqGbWMAAKEgN.jpg   
1824  835152434251116546  https://pbs.twimg.com/media/C5cOtWVWMAEjO5p.jpg   

      img_num          p1   p1_conf  p1_dog                              p2  \
540         1    malamute  0.363257    True                  Siberian_husky   
25          1  guinea_pig  0.996496   False                           skunk   
1465        2    Airedale  0.786089    True                   Irish_terrier   
649         1   Chihuahua  0.203570    True                         doormat   
1824        3       swing  0.967066   False  American_Staffordshire_terrier   

       p2_conf  p2_dog                         p3   p3_conf  p3_dog  
540   0.245862    True                 Eskimo_dog  0.125547    True  
25    0.002402   False                    hamster  0.000461   False  
1465  0.121488    True           Lakeland_terrier  0.014603    True  
649   0.134316   False                toy_terrier  0.084482    True  
1824  0.012731    True  Staffordshire_bullterrier  0.007039    True

In [17]:
sum(df_pred['tweet_id'].duplicated())

0

In [87]:
df_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [88]:
df_pred['img_num'].value_counts()

1    1780
2     198
3      66
4      31
Name: img_num, dtype: int64

### Tidiness
* Not every row is an observation. Each prediction should be its own observation

## Twitter API

In [36]:
df_api

tweet_id retweet_count favorite_count
0     892420643555336193          7465          35355
1     892177421306343426          5541          30600
2     891815181378084864          3666          23025
3     891689557279858688          7641          38655
4     891327558926688256          8234          36926
...                  ...           ...            ...
2326  666049248165822465            39             96
2327  666044226329800704           124            265
2328  666033412701032449            39            109
2329  666029285002620928            41            119
2330  666020888022790149           448           2355

[2331 rows x 3 columns]

In [37]:
df_api.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   tweet_id        2331 non-null   object
 1   retweet_count   2331 non-null   object
 2   favorite_count  2331 non-null   object
dtypes: object(3)
memory usage: 54.8+ KB


In [314]:
len(fails_dict)

25

### Quality
* the 'tweet_id', 'retweet_count' and 'favorite_count' are of type string. Should be integers.
* 25 tweet IDs returned errors
* only 2331 tweets


# Clean

## Twitter Archive

In [56]:
#Copying the original DataFrame into a new one to be cleaned
df_archive_clean = df_archive.copy()

#### Define
* Convert the 'timestamp' and 'retweeted_status_timestamp' columns from string to datetime.

#### Code

In [57]:
df_archive_clean['timestamp'] = pd.to_datetime(df_archive_clean['timestamp'])
df_archive_clean['retweeted_status_timestamp'] = pd.to_datetime(df_archive_clean['retweeted_status_timestamp'])

#### Test

In [58]:
df_archive_clean.dtypes

tweet_id                                    int64
in_reply_to_status_id                     float64
in_reply_to_user_id                       float64
timestamp                     datetime64[ns, UTC]
source                                     object
text                                       object
retweeted_status_id                       float64
retweeted_status_user_id                  float64
retweeted_status_timestamp    datetime64[ns, UTC]
expanded_urls                              object
rating_numerator                            int64
rating_denominator                          int64
name                                       object
doggo                                      object
floofer                                    object
pupper                                     object
puppo                                      object
dtype: object

#### Define

* Combine the dog stages columns ('doggo', 'flooofer', 'pupper', 'puppo') should be one column 'dog_stage' instead.
* for cases where 'doggo' and other value exist. 'doggo' will be chosen as 'doggo' is the older dog and dogs can only grow up. So, 'doggo' will definetly be the more valid value.

#### Code

In [59]:
def dog_stage(row):
    """
    Returns the dog stage based on the values of other columns
    """
    stages = ['doggo','floofer','pupper','puppo']
    for stage in stages:
        if row[stage].lower().strip() == stage:
            return stage
    return 'None'



df_archive_clean['dog_stage'] = df_archive_clean.apply(lambda x: dog_stage(x), axis=1)

#### Test

In [60]:
stages = ['doggo','floofer','pupper','puppo']

for stage in stages:
    print(df_archive_clean[stage].value_counts()[stage])
    print(df_archive_clean['dog_stage'].value_counts()[stage])

97
97
10
9
257
245
30
29


In [61]:
df_archive_clean[['doggo','floofer','pupper','puppo']].value_counts()

doggo  floofer  pupper  puppo
None   None     None    None     1976
                pupper  None      245
doggo  None     None    None       83
None   None     None    puppo      29
doggo  None     pupper  None       12
None   floofer  None    None        9
doggo  floofer  None    None        1
       None     None    puppo       1
dtype: int64

#### Define

* Remove the rows that are retweets.
* Remove the rows that are replies.

#### Code

In [62]:
df_archive_clean = df_archive_clean[df_archive_clean['retweeted_status_id'].isnull()]
df_archive_clean = df_archive_clean[df_archive_clean['in_reply_to_status_id'].isnull()]

#### Test

In [63]:
df_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 18 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   tweet_id                    2097 non-null   int64              
 1   in_reply_to_status_id       0 non-null      float64            
 2   in_reply_to_user_id         0 non-null      float64            
 3   timestamp                   2097 non-null   datetime64[ns, UTC]
 4   source                      2097 non-null   object             
 5   text                        2097 non-null   object             
 6   retweeted_status_id         0 non-null      float64            
 7   retweeted_status_user_id    0 non-null      float64            
 8   retweeted_status_timestamp  0 non-null      datetime64[ns, UTC]
 9   expanded_urls               2094 non-null   object             
 10  rating_numerator            2097 non-null   int64           

#### Define

* Remove the rows with empty 'expanded_urls'. These are the tweets that don't have images.

#### Code

In [64]:
df_archive_clean = df_archive_clean[df_archive_clean['expanded_urls'].notna()]

#### Test

In [65]:
df_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2094 entries, 0 to 2355
Data columns (total 18 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   tweet_id                    2094 non-null   int64              
 1   in_reply_to_status_id       0 non-null      float64            
 2   in_reply_to_user_id         0 non-null      float64            
 3   timestamp                   2094 non-null   datetime64[ns, UTC]
 4   source                      2094 non-null   object             
 5   text                        2094 non-null   object             
 6   retweeted_status_id         0 non-null      float64            
 7   retweeted_status_user_id    0 non-null      float64            
 8   retweeted_status_timestamp  0 non-null      datetime64[ns, UTC]
 9   expanded_urls               2094 non-null   object             
 10  rating_numerator            2094 non-null   int64           

#### Define

* Remove the  columns that are no longer relevant. ('in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp', 'doggo', 'floofer', 'pupper', 'puppo')

#### Code

In [66]:
cols = ['in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp', 'doggo', 'floofer', 'pupper', 'puppo']

df_archive_clean.drop(cols, axis=1, inplace=True)

#### Test

In [67]:
df_archive_clean.columns

Index(['tweet_id', 'timestamp', 'source', 'text', 'expanded_urls',
       'rating_numerator', 'rating_denominator', 'name', 'dog_stage'],
      dtype='object')

#### Define


#### Code

#### Test

## Tweet Image Predictions

In [74]:
# copy into a new DataFrame to be cleaned
df_pred_clean = df_pred.copy()

#### Define

* Convert the dataframe such that each row represents a prediction.

#### Code

In [75]:
# Renaming the dataset columns
cols = ['tweet_id', 'jpg_url', 'img_num', 
       'prediction_1', 'confidence_1', 'breed_1',
       'prediction_2', 'confidence_2', 'breed_2',
       'prediction_3', 'confidence_3', 'breed_3']
df_pred_clean.columns = cols

df_pred_clean = pd.wide_to_long(df_pred_clean, stubnames=['prediction', 'confidence', 'breed'], 
    i=['tweet_id', 'jpg_url', 'img_num'], j='prediction_level', sep="_").reset_index()

#### Test

In [77]:
df_pred_clean.describe()

tweet_id      img_num  prediction_level    confidence
count  6.225000e+03  6225.000000       6225.000000  6.225000e+03
mean   7.384514e+17     1.203855          2.000000  2.631537e-01
std    6.784113e+16     0.561785          0.816562  2.908324e-01
min    6.660209e+17     1.000000          1.000000  1.740170e-10
25%    6.764706e+17     1.000000          1.000000  5.123350e-02
50%    7.119988e+17     1.000000          2.000000  1.351790e-01
75%    7.932110e+17     1.000000          3.000000  3.796240e-01
max    8.924206e+17     4.000000          3.000000  1.000000e+00

## Twitter API

In [79]:
df_api_clean = df_api.copy()

#### Define

* Convert the 'tweet_id', 'retweet_count' and 'favorite_count' columns to integer instead of string.

#### Code

In [80]:
df_api_clean['tweet_id'] = df_api_clean['tweet_id'].astype(int)
df_api_clean['retweet_count'] = df_api_clean['retweet_count'].astype(int)
df_api_clean['favorite_count'] = df_api_clean['favorite_count'].astype(int)

#### Test

In [81]:
df_api_clean.dtypes

tweet_id          int64
retweet_count     int64
favorite_count    int64
dtype: object

## Merging Data

#### Define

* Merge the `df_archive_clean` and the `df_api_clean` based on the 'tweet_id'

#### Code

In [82]:
df_merged = pd.merge(df_archive_clean, df_api_clean, on='tweet_id')

#### Test

In [86]:
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2087 entries, 0 to 2086
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            2087 non-null   int64              
 1   timestamp           2087 non-null   datetime64[ns, UTC]
 2   source              2087 non-null   object             
 3   text                2087 non-null   object             
 4   expanded_urls       2087 non-null   object             
 5   rating_numerator    2087 non-null   int64              
 6   rating_denominator  2087 non-null   int64              
 7   name                2087 non-null   object             
 8   dog_stage           2087 non-null   object             
 9   retweet_count       2087 non-null   int64              
 10  favorite_count      2087 non-null   int64              
dtypes: datetime64[ns, UTC](1), int64(5), object(5)
memory usage: 195.7+ KB


## Storing Data

export the df_merged file to a new file 'twitter_archive_clean.csv'

In [87]:
df_merged.to_csv('twitter_archive_clean.csv')

#### Test

#### Define


#### Code

#### Test

#### Define


#### Code

#### Test